{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ff4243a0-ef2e-400d-84bd-9b0e67d0d670",
   "metadata": {},
   "source": [
    "# Example: Create a cell type reference gene expression file from a scRNA-seq atlas with cell type annotation. \n",
    "- Using Human colon cancer atlas as example.\n",
    "- Data downloaded from GSE178341, with h5 and cell type labels. \n",
    "## We already know cutoffs for good quality so Step0 can be skipped:\n",
    "- MIN_GENES = 100\n",
    "- MIN_COUNTS = 100\n",
    "- MIN_CELLS = 3\n",
    "- MAX_MITO_PERCENT=27.77"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "156cd0c4-7ce0-4a13-aeb7-4fe2074243ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "# add STHD path\n",
    "sys.path.append('../')\n",
    "\n",
    "# (pip version to be published)\n",
    "# pip install STHD"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22bc35fa-91e5-4ec2-b1fa-fd252ebadec8",
   "metadata": {},
   "source": [
    "# Step 0 (Optional): Find Good QC cutoffs  \n",
    "## Load scRNA reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a3f32e32-a147-44ed-be46-b4b4730fee60",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/hpc/group/yizhanglab/yiz/venv/stvit-cpu/lib64/python3.9/site-packages/anndata/_core/anndata.py:1908: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/hpc/group/yizhanglab/yiz/venv/stvit-cpu/lib64/python3.9/site-packages/anndata/_core/anndata.py:1908: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    }
   ],
   "source": [
    "crc_dir = '../testdata/crc_scrna/'\n",
    "adata = sc.read_10x_h5('../testdata/crc_scrna/GSE178341_crc10x_full_c295v4_submit.h5')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98a3dd1d-ab0a-4d62-9d34-b081b51aced3",
   "metadata": {},
   "source": [
    "## Load cell type names and add to adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "50c26be1-0c56-43af-b90d-8d69ae27af8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_global_celltype(crc_dir):\n",
    "    '''\n",
    "    For colon cancer atlas\n",
    "    '''\n",
    "    cell_type_tsne_type_file = 'crc10x_tSNE_cl_global.tsv'\n",
    "    tmp = pd.read_table( os.path.join(crc_dir, cell_type_tsne_type_file), skiprows=1, index_col = 0) # same cell type and number\n",
    "    return(tmp)\n",
    "\n",
    "celltypes = load_global_celltype(crc_dir)\n",
    "# qc\n",
    "celltypes = celltypes.loc[ celltypes.index.intersection(adata.obs.index) ]\n",
    "celltypes_group1 = celltypes.groupby('group.1').count().index.tolist()\n",
    "celltypes_group = celltypes.groupby('group').count().index.tolist()\n",
    "\n",
    "adata.obsm['X_tsne'] = celltypes[['numeric','numeric.1']].loc[adata.obs.index].values\n",
    "adata.obs['group']=celltypes['group'].loc[adata.obs.index].values\n",
    "adata.obs['group.1']=celltypes['group.1'].loc[adata.obs.index].values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1d5502c-796e-493b-b146-70c278fd022c",
   "metadata": {},
   "source": [
    "## Optional plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4e39b49-acb3-4583-a5f7-aab4fe49e1f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata,color='group.1')\n",
    "sc.pl.tsne(adata,color='group')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1b8302b-3233-4b55-a260-8bf8c95211ba",
   "metadata": {},
   "source": [
    "## Finding QC criterion: Optional for first time QC: QC and mitochondrial control. can skip if you know the QC cutoffs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "abcfa8c0-3a54-4602-be64-4f80f6c93335",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "MIN_GENES = 100\n",
    "MIN_COUNTS = 100\n",
    "MIN_CELLS = 3\n",
    "random_state = 42\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58cf89e8-1335-43b9-8e9b-e58fb2b39782",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var_names_make_unique()\n",
    "adata.var['mt'] = adata.var_names.str.startswith('MT-') | adata.var_names.str.startswith('mt-')\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "sc.pl.violin(\n",
    "    adata,\n",
    "    [\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\n",
    "    jitter=0.4,\n",
    "    multi_panel=True,\n",
    ")\n",
    "sc.pl.scatter(adata, \"total_counts\", \"n_genes_by_counts\", color=\"pct_counts_mt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3af8be8-e0da-48bd-8b78-c0b98de16106",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "sc.pl.tsne(adata,color='pct_counts_mt')\n",
    "#plt.savefig('../analysis/figure/crc95_ref_qc_pct_counts_mt.pdf', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de98df18-1ed9-4777-b061-772e2f51d0d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "sns.displot(adata.obs['pct_counts_mt'].values)\n",
    "plt.axvline(27.7, color='pink')\n",
    "plt.xlabel('pct_counts_mt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a819024b-70b3-4510-b02c-85e7a4c3c46f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mito cutoff option: 27.771158613939527\n"
     ]
    }
   ],
   "source": [
    "# either\n",
    "\n",
    "MAX_MITO_PERCENT = adata.obs['pct_counts_mt'].mean() + adata.obs['pct_counts_mt'].std()\n",
    "print(f'Mito cutoff option: {MAX_MITO_PERCENT}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b5a3305b-e03d-4177-9c9c-1edb33bd2d2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pre mt filtering  370115\n",
      "post mt filtering  303358\n"
     ]
    }
   ],
   "source": [
    "sc.pp.filter_cells(adata, min_genes = MIN_GENES)  \n",
    "sc.pp.filter_cells(adata, min_counts = MIN_COUNTS )   \n",
    "sc.pp.filter_genes(adata, min_cells = MIN_CELLS)\n",
    "\n",
    "print('pre mt filtering ', adata.shape[0])\n",
    "adata = adata[adata.obs.pct_counts_mt < MAX_MITO_PERCENT , :]\n",
    "print('post mt filtering ', adata.shape[0])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd282225-9141-4592-af32-b549d030c94d",
   "metadata": {},
   "source": [
    "# Step 1: Read in the reference scRNA and QC "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8d459add-5c80-4ce6-aa3f-05ef8fbd638f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "09b5a3c1-8ae4-484e-99c8-5051a3d8ccaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/hpc/group/yizhanglab/yiz/venv/stvit-cpu/lib64/python3.9/site-packages/anndata/_core/anndata.py:1908: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/hpc/group/yizhanglab/yiz/venv/stvit-cpu/lib64/python3.9/site-packages/anndata/_core/anndata.py:1908: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    }
   ],
   "source": [
    "crc_dir = '../testdata/crc_scrna/'\n",
    "adata = sc.read_10x_h5('../testdata/crc_scrna/GSE178341_crc10x_full_c295v4_submit.h5')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10f8faa4-a338-4d73-b195-5eed6c9bf249",
   "metadata": {},
   "source": [
    "## Get cell type and load to raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "071ca509-d0c0-488e-b8b4-8f7aaaa706ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_global_celltype(crc_dir):\n",
    "    '''\n",
    "    For colon cancer atlas\n",
    "    '''\n",
    "    cell_type_tsne_type_file = 'crc10x_tSNE_cl_global.tsv'\n",
    "    tmp = pd.read_table( os.path.join(crc_dir, cell_type_tsne_type_file), skiprows=1, index_col = 0) # same cell type and number\n",
    "    return(tmp)\n",
    "\n",
    "celltypes = load_global_celltype(crc_dir)\n",
    "# qc\n",
    "celltypes = celltypes.loc[ celltypes.index.intersection(adata.obs.index) ]\n",
    "celltypes_group1 = celltypes.groupby('group.1').count().index.tolist()\n",
    "celltypes_group = celltypes.groupby('group').count().index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9ae12f90-3de7-4e91-98c8-6e859906d2cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm['X_tsne'] = celltypes[['numeric','numeric.1']].loc[adata.obs.index].values\n",
    "adata.obs['group']=celltypes['group'].loc[adata.obs.index].values\n",
    "adata.obs['group.1']=celltypes['group.1'].loc[adata.obs.index].values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b950a722-0463-465c-ab5f-ed27624b3753",
   "metadata": {},
   "source": [
    "## KEEP the raw counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e024cebf-9e1c-4067-9bbe-6f91083a63de",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/hpc/group/yizhanglab/yiz/venv/stvit-cpu/lib64/python3.9/site-packages/anndata/utils.py:260: UserWarning: Suffix used (-[0-9]+) to deduplicate index values may make index values difficult to interpret. There values with a similar suffixes in the index. Consider using a different delimiter by passing `join={delimiter}`Example key collisions generated by the make_index_unique algorithm: ['SNORD116-1', 'SNORD116-2', 'SNORD116-3', 'SNORD116-5', 'SNORD116-6']\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "adata.var_names_make_unique()\n",
    "if('counts') not in adata.layers.keys():\n",
    "    adata.layers[\"counts\"] = adata.X.copy()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d79448d3-e259-4ff6-9c0c-6f74a1beb02a",
   "metadata": {},
   "source": [
    "### QC and Mitochondrial control: we use the thresholds from observations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "60d8076b-4753-477e-a812-613905a48909",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pre mt filtering  370115\n",
      "post mt filtering  303350\n"
     ]
    }
   ],
   "source": [
    "\n",
    "adata.var['mt'] = adata.var_names.str.startswith('MT-') | adata.var_names.str.startswith('mt-')\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "MIN_GENES = 100\n",
    "MIN_COUNTS = 100\n",
    "MIN_CELLS = 3\n",
    "random_state = 42\n",
    "\n",
    "sc.pp.filter_cells(adata, min_genes = MIN_GENES)  \n",
    "sc.pp.filter_cells(adata, min_counts = MIN_COUNTS )   \n",
    "sc.pp.filter_genes(adata, min_cells = MIN_CELLS)\n",
    "\n",
    "MAX_MITO_PERCENT=27.77\n",
    "\n",
    "print('pre mt filtering ', adata.shape[0])\n",
    "adata = adata[adata.obs.pct_counts_mt < MAX_MITO_PERCENT , :]\n",
    "print('post mt filtering ', adata.shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d3e5ea50-65f3-4fa1-a855-fb2b4d91c842",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12301.415025548047"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['n_counts'].values.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6648a6d4-f3bb-4503-95f4-f1edb789f4f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sc.pl.tsne(adata,color='group', save='crc98_ref_celltype_umap.pdf')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "404b6f17-0cde-4b0a-ad7b-91f0b815fd88",
   "metadata": {},
   "source": [
    "# Step 2: Convert gene counts into normalized gene expression per-cell type for gene selection\n",
    "- proportion of genes sum to 1 for a cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5dcb63bb-0920-4c3c-a012-73f1a3cd8be1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import STHD.refscrna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a91cb107-6950-459b-81f0-3d522fe6c3c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Log] 300 cells for cell type cM03 (DC1)\n",
      "dimension of cell by gene count (300, 31873)\n",
      "[Log] 11362 cells for cell type cB2 (B GC-like)\n",
      "dimension of cell by gene count (11362, 31873)\n",
      "[Log] 722 cells for cell type cS05 (Endo venous)\n",
      "dimension of cell by gene count (722, 31873)\n",
      "[Log] 15824 cells for cell type cP2 (Plasma IgG)\n",
      "dimension of cell by gene count (15824, 31873)\n",
      "[Log] 485 cells for cell type cTNI12 (CD8+ IL7R+)\n",
      "dimension of cell by gene count (485, 31873)\n",
      "[Log] 386 cells for cell type cS10 (Endo tip cells)\n",
      "dimension of cell by gene count (386, 31873)\n",
      "[Log] 94 cells for cell type cM08 (AS-DC)\n",
      "dimension of cell by gene count (94, 31873)\n",
      "[Log] 1299 cells for cell type Tumor cE05 (Enterocyte 2)\n",
      "dimension of cell by gene count (1299, 31873)\n",
      "[Log] 518 cells for cell type cS21 (Fibro stem cell niche)\n",
      "dimension of cell by gene count (518, 31873)\n",
      "[Log] 788 cells for cell type cTNI22 (cTNI22)\n",
      "dimension of cell by gene count (788, 31873)\n",
      "[Log] 4516 cells for cell type cTNI11 (CD8+GZMK+)\n",
      "dimension of cell by gene count (4516, 31873)\n",
      "[Log] 4270 cells for cell type cTNI02 (CD4+ IL7R+SELL+)\n",
      "dimension of cell by gene count (4270, 31873)\n",
      "[Log] 1234 cells for cell type cTNI23 (NK CD16A+)\n",
      "dimension of cell by gene count (1234, 31873)\n",
      "[Log] 1247 cells for cell type cS13 (Endo venous-like)\n",
      "dimension of cell by gene count (1247, 31873)\n",
      "[Log] 13765 cells for cell type cM01 (Monocyte)\n",
      "dimension of cell by gene count (13765, 31873)\n",
      "[Log] 349 cells for cell type cS19 (Pericyte)\n",
      "dimension of cell by gene count (349, 31873)\n",
      "[Log] 95 cells for cell type cE11 (Enteroendocrine)\n",
      "dimension of cell by gene count (95, 31873)\n",
      "[Log] 282 cells for cell type cS06 (Endo lymphatic)\n",
      "dimension of cell by gene count (282, 31873)\n",
      "[Log] 337 cells for cell type cS17 (Pericyte)\n",
      "dimension of cell by gene count (337, 31873)\n",
      "[Log] 6782 cells for cell type Tumor cE02 (Stem/TA-like/Immature Goblet)\n",
      "dimension of cell by gene count (6782, 31873)\n",
      "[Log] 441 cells for cell type cS08 (Endo arterial-like)\n",
      "dimension of cell by gene count (441, 31873)\n",
      "[Log] 1876 cells for cell type cTNI03 (CD4+ IL7R+HSP+)\n",
      "dimension of cell by gene count (1876, 31873)\n",
      "[Log] 1066 cells for cell type cM05 (DC2 C1Q+)\n",
      "dimension of cell by gene count (1066, 31873)\n",
      "[Log] 3917 cells for cell type cE05 (Enterocyte 2)\n",
      "dimension of cell by gene count (3917, 31873)\n",
      "[Log] 1775 cells for cell type cTNI09 (CD4+ Treg prolif)\n",
      "dimension of cell by gene count (1775, 31873)\n",
      "[Log] 437 cells for cell type cS01 (Endo arterial)\n",
      "dimension of cell by gene count (437, 31873)\n",
      "[Log] 673 cells for cell type cS27 (CXCL14+ CAF)\n",
      "dimension of cell by gene count (673, 31873)\n",
      "[Log] 5421 cells for cell type cTNI17 (gd-like T)\n",
      "dimension of cell by gene count (5421, 31873)\n",
      "[Log] 1090 cells for cell type cM04 (DC2)\n",
      "dimension of cell by gene count (1090, 31873)\n",
      "[Log] 1008 cells for cell type Tumor cE10 (Tuft)\n",
      "dimension of cell by gene count (1008, 31873)\n",
      "[Log] 412 cells for cell type cS03 (Endo capillary)\n",
      "dimension of cell by gene count (412, 31873)\n",
      "[Log] 9530 cells for cell type cTNI08 (CD4+ Treg)\n",
      "dimension of cell by gene count (9530, 31873)\n",
      "[Log] 254 cells for cell type Tumor cE11 (Enteroendocrine)\n",
      "dimension of cell by gene count (254, 31873)\n",
      "[Log] 194 cells for cell type cS04 (Endo)\n",
      "dimension of cell by gene count (194, 31873)\n",
      "[Log] 660 cells for cell type cTNI21 (PLZF+ T prolif)\n",
      "dimension of cell by gene count (660, 31873)\n",
      "[Log] 216 cells for cell type cS26 (Myofibro)\n",
      "dimension of cell by gene count (216, 31873)\n",
      "[Log] 1871 cells for cell type cM10 (Granulocyte)\n",
      "dimension of cell by gene count (1871, 31873)\n",
      "[Log] 314 cells for cell type cE07 (Goblet/Enterocyte)\n",
      "dimension of cell by gene count (314, 31873)\n",
      "[Log] 38 cells for cell type cS14 (Endo)\n",
      "dimension of cell by gene count (38, 31873)\n",
      "[Log] 6182 cells for cell type cB3 (B CD40+ GC-like)\n",
      "dimension of cell by gene count (6182, 31873)\n",
      "[Log] 44 cells for cell type cS20 (Pericyte prolif)\n",
      "dimension of cell by gene count (44, 31873)\n",
      "[Log] 797 cells for cell type cTNI26 (ILC3)\n",
      "dimension of cell by gene count (797, 31873)\n",
      "[Log] 14 cells for cell type Tumor cE07 (Goblet/Enterocyte)\n",
      "dimension of cell by gene count (14, 31873)\n",
      "[Log] 6330 cells for cell type cTNI14 (CD8+ CXCL13+)\n",
      "dimension of cell by gene count (6330, 31873)\n",
      "[Log] 2920 cells for cell type Tumor cE06 (Immature Goblet)\n",
      "dimension of cell by gene count (2920, 31873)\n",
      "[Log] 217 cells for cell type cS23 (Fibro BMP-producing)\n",
      "dimension of cell by gene count (217, 31873)\n",
      "[Log] 1999 cells for cell type cTNI13 (CD8+ T IL17+)\n",
      "dimension of cell by gene count (1999, 31873)\n",
      "[Log] 194 cells for cell type cS11 (Endo proif)\n",
      "dimension of cell by gene count (194, 31873)\n",
      "[Log] 2095 cells for cell type cTNI07 (CD4+ CXCL13+)\n",
      "dimension of cell by gene count (2095, 31873)\n",
      "[Log] 439 cells for cell type cS25 (Fibro CCL8+)\n",
      "dimension of cell by gene count (439, 31873)\n",
      "[Log] 19434 cells for cell type cM02 (Macrophage-like)\n",
      "dimension of cell by gene count (19434, 31873)\n",
      "[Log] 93 cells for cell type cS07 (Endo capillary-like)\n",
      "dimension of cell by gene count (93, 31873)\n",
      "[Log] 361 cells for cell type cS29 (MMP3+ CAF)\n",
      "dimension of cell by gene count (361, 31873)\n",
      "[Log] 181 cells for cell type cS16 (Pericyte)\n",
      "dimension of cell by gene count (181, 31873)\n",
      "[Log] 1551 cells for cell type cM09 (mregDC)\n",
      "dimension of cell by gene count (1551, 31873)\n",
      "[Log] 497 cells for cell type cM06 (DC IL22RA2)\n",
      "dimension of cell by gene count (497, 31873)\n",
      "[Log] 2222 cells for cell type cE08 (Goblet)\n",
      "dimension of cell by gene count (2222, 31873)\n",
      "[Log] 717 cells for cell type cS09 (Endo)\n",
      "dimension of cell by gene count (717, 31873)\n",
      "[Log] 223 cells for cell type cS18 (Pericyte)\n",
      "dimension of cell by gene count (223, 31873)\n",
      "[Log] 2885 cells for cell type cTNI01 (CD4+ IL7R+)\n",
      "dimension of cell by gene count (2885, 31873)\n",
      "[Log] 639 cells for cell type cS28 (GREM1+ CAF)\n",
      "dimension of cell by gene count (639, 31873)\n",
      "[Log] 228 cells for cell type cS30 (CAF CCL8 Fibro-like)\n",
      "dimension of cell by gene count (228, 31873)\n",
      "[Log] 2255 cells for cell type cTNI16 (CD8+ CXCL13+ prolif)\n",
      "dimension of cell by gene count (2255, 31873)\n",
      "[Log] 20107 cells for cell type cP1 (Plasma IgA)\n",
      "dimension of cell by gene count (20107, 31873)\n",
      "[Log] 29726 cells for cell type Tumor cE03 (Stem/TA-like prolif)\n",
      "dimension of cell by gene count (29726, 31873)\n",
      "[Log] 5966 cells for cell type cTNI04 (CD4+ IL7R+CCL5+)\n",
      "dimension of cell by gene count (5966, 31873)\n",
      "[Log] 2832 cells for cell type cE06 (Immature Goblet)\n",
      "dimension of cell by gene count (2832, 31873)\n",
      "[Log] 1200 cells for cell type cS02 (Endo capillary)\n",
      "dimension of cell by gene count (1200, 31873)\n",
      "[Log] 383 cells for cell type cS24 (Fibro BMP-producing)\n",
      "dimension of cell by gene count (383, 31873)\n",
      "[Log] 473 cells for cell type Tumor cE08 (Goblet)\n",
      "dimension of cell by gene count (473, 31873)\n",
      "[Log] 4141 cells for cell type cE09 (Best4)\n",
      "dimension of cell by gene count (4141, 31873)\n",
      "[Log] 840 cells for cell type cM07 (pDC)\n",
      "dimension of cell by gene count (840, 31873)\n",
      "[Log] 1477 cells for cell type cTNI06 (CD4+ TFH)\n",
      "dimension of cell by gene count (1477, 31873)\n",
      "[Log] 2803 cells for cell type cTNI05 (CD4+ IL17+)\n",
      "dimension of cell by gene count (2803, 31873)\n",
      "[Log] 3325 cells for cell type cE03 (Stem/TA-like prolif)\n",
      "dimension of cell by gene count (3325, 31873)\n",
      "[Log] 457 cells for cell type cS31 (CAF stem niche Fibro-like)\n",
      "dimension of cell by gene count (457, 31873)\n",
      "[Log] 627 cells for cell type cS12 (Endo)\n",
      "dimension of cell by gene count (627, 31873)\n",
      "[Log] 3671 cells for cell type cMA01 (Mast)\n",
      "dimension of cell by gene count (3671, 31873)\n",
      "[Log] 254 cells for cell type cS33 (Schwann)\n",
      "dimension of cell by gene count (254, 31873)\n",
      "[Log] 1117 cells for cell type cTNI24 (NK GZMK+)\n",
      "dimension of cell by gene count (1117, 31873)\n",
      "[Log] 643 cells for cell type cTNI25 (NK XCL1+)\n",
      "dimension of cell by gene count (643, 31873)\n",
      "[Log] 2682 cells for cell type cTNI10 (CD8+ IL7R+)\n",
      "dimension of cell by gene count (2682, 31873)\n",
      "[Log] 936 cells for cell type cE10 (Tuft)\n",
      "dimension of cell by gene count (936, 31873)\n",
      "[Log] 252 cells for cell type cS15 (Pericyte)\n",
      "dimension of cell by gene count (252, 31873)\n",
      "[Log] 2708 cells for cell type cE02 (Stem/TA-like/Immature Goblet)\n",
      "dimension of cell by gene count (2708, 31873)\n",
      "[Log] 800 cells for cell type cS32 (Smooth Muscle)\n",
      "dimension of cell by gene count (800, 31873)\n",
      "[Log] 369 cells for cell type cTNI19 (gd-like T prolif)\n",
      "dimension of cell by gene count (369, 31873)\n",
      "[Log] 203 cells for cell type Tumor cE09 (Best4)\n",
      "dimension of cell by gene count (203, 31873)\n",
      "[Log] 688 cells for cell type cS22 (Fibro stem cell niche)\n",
      "dimension of cell by gene count (688, 31873)\n",
      "[Log] 3920 cells for cell type cTNI20 (PLZF+ T)\n",
      "dimension of cell by gene count (3920, 31873)\n",
      "[Log] 36490 cells for cell type Tumor cE01 (Stem/TA-like)\n",
      "dimension of cell by gene count (36490, 31873)\n",
      "[Log] 3878 cells for cell type cTNI15 (CD8+ CXCL13+ HSP+)\n",
      "dimension of cell by gene count (3878, 31873)\n",
      "[Log] 960 cells for cell type Tumor cE04 (Enterocyte 1)\n",
      "dimension of cell by gene count (960, 31873)\n",
      "[Log] 2003 cells for cell type cE04 (Enterocyte 1)\n",
      "dimension of cell by gene count (2003, 31873)\n",
      "[Log] 3178 cells for cell type cTNI18 (gd-like T PDCD1+)\n",
      "dimension of cell by gene count (3178, 31873)\n",
      "[Log] 6939 cells for cell type cB1 (B IGD+IgM+)\n",
      "dimension of cell by gene count (6939, 31873)\n",
      "[Log] 7886 cells for cell type cE01 (Stem/TA-like)\n",
      "dimension of cell by gene count (7886, 31873)\n",
      "[Log] 1051 cells for cell type cP3 (Plasma IgG prolif)\n",
      "dimension of cell by gene count (1051, 31873)\n"
     ]
    }
   ],
   "source": [
    "from STHD.refscrna import gene_lambda_by_ct\n",
    "genemeanpd, genemeanpdfc = gene_lambda_by_ct(adata, ctcol = 'group')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50166420-adeb-4501-aa48-7d6850828eb6",
   "metadata": {},
   "source": [
    "# Step 3: Select some cell type-informative genes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba831ded-62d0-4bb0-8c27-97391e7850c8",
   "metadata": {},
   "source": [
    "- Min gene lambda 0.000125 , logFC 0.5. \n",
    "- This is example cutoff refers to RCTD STYLE: \"Using the estimated cell type expression profiles ðœ‡Ë†ð‘˜,ð‘—, we select differentially expressed genes that will be informative when estimating cell type proportions. For each cell type in the scRNA-seq reference, we select genes with minimum average expression above .0625 counts per 500 and at least 0.5 log-fold-change compared to the average expression across all cell types. \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ef82566d-b3d7-497b-a09c-4d0aec9760e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31734 genes after removing MT- and Ribo\n",
      "4618 genes after further removing RCTD threshold gene expression 0.0125\n",
      "4618 genes after further removing RCTD logfc log2 0.5\n",
      "(4618, 98)\n"
     ]
    }
   ],
   "source": [
    "from STHD.refscrna import select_ct_informative_genes_basic\n",
    "genemeanpd_filtered = select_ct_informative_genes_basic(genemeanpd, genemeanpdfc, exprcut = 0.000125, fccut = 2**0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "97a2e026-9245-44b3-9fc6-b5a6eb1b6e3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'LEFTY1,ALDH1B1,CDCA7,OLFM4,STMN1,HMGB2,MKI67,CLCA4,AQP8,TMIGD1,GUCA2B,SULT1A2,SLC26A3,SLC26A2,CA1,SPINK1,CLCA1,ITLN1,LRRC26,RETNLB,SPINK4,FCGBP,MUC2,ZG16,BCAS1,ENTPD8,MUC1,TFF1,S100P,NOTCH2NL,BEST4,OTOP2,CA7,AZGP1,HCK,LRMP,SH2D6,HPGDS,PTGS1,CHGA,CRYBA2,SCGN,PYY,SCT,INSL5'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check a few gene markers to see whether they are in.\n",
    "colon_epithelial_marker = ['LEFTY1','EPHB3','LGR5','ALDH1B1','CDCA7','OLFM4','STMN1','HMGB2','MKI67','CLCA4','AQP8','TMIGD1','GUCA2B','SULT1A2','SLC26A3','SLC26A2','CA1',\n",
    "           'SPINK1','CLCA1','ITLN1','LRRC26','RETNLB','SPINK4','FCGBP','MUC2','ZG16','BCAS1','ENTPD8','MUC1','TFF1','S100P','NOTCH2NL','OTOP3','REN','BEST4','OTOP2','CA7',\n",
    "          'AZGP1','HCK','LRMP','SH2D6','HPGDS','PTGS1','CHGA','CRYBA2','SCGN','PYY','SCT','INSL5'] \n",
    "','.join([t for t in colon_epithelial_marker  if t in genemeanpd_filtered.index])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0f7004d-db5b-4dd9-9cb9-fa5c68c06d2a",
   "metadata": {},
   "source": [
    "## Now, Re-calculate the gene labmdas for the selected celltype-informative genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "071804c1-3e2f-4f08-abd9-5cc219663fe1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Log] 300 cells for cell type cM03 (DC1)\n",
      "dimension of cell by gene count (300, 4618)\n",
      "[Log] 11362 cells for cell type cB2 (B GC-like)\n",
      "dimension of cell by gene count (11362, 4618)\n",
      "[Log] 722 cells for cell type cS05 (Endo venous)\n",
      "dimension of cell by gene count (722, 4618)\n",
      "[Log] 15824 cells for cell type cP2 (Plasma IgG)\n",
      "dimension of cell by gene count (15824, 4618)\n",
      "[Log] 485 cells for cell type cTNI12 (CD8+ IL7R+)\n",
      "dimension of cell by gene count (485, 4618)\n",
      "[Log] 386 cells for cell type cS10 (Endo tip cells)\n",
      "dimension of cell by gene count (386, 4618)\n",
      "[Log] 94 cells for cell type cM08 (AS-DC)\n",
      "dimension of cell by gene count (94, 4618)\n",
      "[Log] 1299 cells for cell type Tumor cE05 (Enterocyte 2)\n",
      "dimension of cell by gene count (1299, 4618)\n",
      "[Log] 518 cells for cell type cS21 (Fibro stem cell niche)\n",
      "dimension of cell by gene count (518, 4618)\n",
      "[Log] 788 cells for cell type cTNI22 (cTNI22)\n",
      "dimension of cell by gene count (788, 4618)\n",
      "[Log] 4516 cells for cell type cTNI11 (CD8+GZMK+)\n",
      "dimension of cell by gene count (4516, 4618)\n",
      "[Log] 4270 cells for cell type cTNI02 (CD4+ IL7R+SELL+)\n",
      "dimension of cell by gene count (4270, 4618)\n",
      "[Log] 1234 cells for cell type cTNI23 (NK CD16A+)\n",
      "dimension of cell by gene count (1234, 4618)\n",
      "[Log] 1247 cells for cell type cS13 (Endo venous-like)\n",
      "dimension of cell by gene count (1247, 4618)\n",
      "[Log] 13765 cells for cell type cM01 (Monocyte)\n",
      "dimension of cell by gene count (13765, 4618)\n",
      "[Log] 349 cells for cell type cS19 (Pericyte)\n",
      "dimension of cell by gene count (349, 4618)\n",
      "[Log] 95 cells for cell type cE11 (Enteroendocrine)\n",
      "dimension of cell by gene count (95, 4618)\n",
      "[Log] 282 cells for cell type cS06 (Endo lymphatic)\n",
      "dimension of cell by gene count (282, 4618)\n",
      "[Log] 337 cells for cell type cS17 (Pericyte)\n",
      "dimension of cell by gene count (337, 4618)\n",
      "[Log] 6782 cells for cell type Tumor cE02 (Stem/TA-like/Immature Goblet)\n",
      "dimension of cell by gene count (6782, 4618)\n",
      "[Log] 441 cells for cell type cS08 (Endo arterial-like)\n",
      "dimension of cell by gene count (441, 4618)\n",
      "[Log] 1876 cells for cell type cTNI03 (CD4+ IL7R+HSP+)\n",
      "dimension of cell by gene count (1876, 4618)\n",
      "[Log] 1066 cells for cell type cM05 (DC2 C1Q+)\n",
      "dimension of cell by gene count (1066, 4618)\n",
      "[Log] 3917 cells for cell type cE05 (Enterocyte 2)\n",
      "dimension of cell by gene count (3917, 4618)\n",
      "[Log] 1775 cells for cell type cTNI09 (CD4+ Treg prolif)\n",
      "dimension of cell by gene count (1775, 4618)\n",
      "[Log] 437 cells for cell type cS01 (Endo arterial)\n",
      "dimension of cell by gene count (437, 4618)\n",
      "[Log] 673 cells for cell type cS27 (CXCL14+ CAF)\n",
      "dimension of cell by gene count (673, 4618)\n",
      "[Log] 5421 cells for cell type cTNI17 (gd-like T)\n",
      "dimension of cell by gene count (5421, 4618)\n",
      "[Log] 1090 cells for cell type cM04 (DC2)\n",
      "dimension of cell by gene count (1090, 4618)\n",
      "[Log] 1008 cells for cell type Tumor cE10 (Tuft)\n",
      "dimension of cell by gene count (1008, 4618)\n",
      "[Log] 412 cells for cell type cS03 (Endo capillary)\n",
      "dimension of cell by gene count (412, 4618)\n",
      "[Log] 9530 cells for cell type cTNI08 (CD4+ Treg)\n",
      "dimension of cell by gene count (9530, 4618)\n",
      "[Log] 254 cells for cell type Tumor cE11 (Enteroendocrine)\n",
      "dimension of cell by gene count (254, 4618)\n",
      "[Log] 194 cells for cell type cS04 (Endo)\n",
      "dimension of cell by gene count (194, 4618)\n",
      "[Log] 660 cells for cell type cTNI21 (PLZF+ T prolif)\n",
      "dimension of cell by gene count (660, 4618)\n",
      "[Log] 216 cells for cell type cS26 (Myofibro)\n",
      "dimension of cell by gene count (216, 4618)\n",
      "[Log] 1871 cells for cell type cM10 (Granulocyte)\n",
      "dimension of cell by gene count (1871, 4618)\n",
      "[Log] 314 cells for cell type cE07 (Goblet/Enterocyte)\n",
      "dimension of cell by gene count (314, 4618)\n",
      "[Log] 38 cells for cell type cS14 (Endo)\n",
      "dimension of cell by gene count (38, 4618)\n",
      "[Log] 6182 cells for cell type cB3 (B CD40+ GC-like)\n",
      "dimension of cell by gene count (6182, 4618)\n",
      "[Log] 44 cells for cell type cS20 (Pericyte prolif)\n",
      "dimension of cell by gene count (44, 4618)\n",
      "[Log] 797 cells for cell type cTNI26 (ILC3)\n",
      "dimension of cell by gene count (797, 4618)\n",
      "[Log] 14 cells for cell type Tumor cE07 (Goblet/Enterocyte)\n",
      "dimension of cell by gene count (14, 4618)\n",
      "[Log] 6330 cells for cell type cTNI14 (CD8+ CXCL13+)\n",
      "dimension of cell by gene count (6330, 4618)\n",
      "[Log] 2920 cells for cell type Tumor cE06 (Immature Goblet)\n",
      "dimension of cell by gene count (2920, 4618)\n",
      "[Log] 217 cells for cell type cS23 (Fibro BMP-producing)\n",
      "dimension of cell by gene count (217, 4618)\n",
      "[Log] 1999 cells for cell type cTNI13 (CD8+ T IL17+)\n",
      "dimension of cell by gene count (1999, 4618)\n",
      "[Log] 194 cells for cell type cS11 (Endo proif)\n",
      "dimension of cell by gene count (194, 4618)\n",
      "[Log] 2095 cells for cell type cTNI07 (CD4+ CXCL13+)\n",
      "dimension of cell by gene count (2095, 4618)\n",
      "[Log] 439 cells for cell type cS25 (Fibro CCL8+)\n",
      "dimension of cell by gene count (439, 4618)\n",
      "[Log] 19434 cells for cell type cM02 (Macrophage-like)\n",
      "dimension of cell by gene count (19434, 4618)\n",
      "[Log] 93 cells for cell type cS07 (Endo capillary-like)\n",
      "dimension of cell by gene count (93, 4618)\n",
      "[Log] 361 cells for cell type cS29 (MMP3+ CAF)\n",
      "dimension of cell by gene count (361, 4618)\n",
      "[Log] 181 cells for cell type cS16 (Pericyte)\n",
      "dimension of cell by gene count (181, 4618)\n",
      "[Log] 1551 cells for cell type cM09 (mregDC)\n",
      "dimension of cell by gene count (1551, 4618)\n",
      "[Log] 497 cells for cell type cM06 (DC IL22RA2)\n",
      "dimension of cell by gene count (497, 4618)\n",
      "[Log] 2222 cells for cell type cE08 (Goblet)\n",
      "dimension of cell by gene count (2222, 4618)\n",
      "[Log] 717 cells for cell type cS09 (Endo)\n",
      "dimension of cell by gene count (717, 4618)\n",
      "[Log] 223 cells for cell type cS18 (Pericyte)\n",
      "dimension of cell by gene count (223, 4618)\n",
      "[Log] 2885 cells for cell type cTNI01 (CD4+ IL7R+)\n",
      "dimension of cell by gene count (2885, 4618)\n",
      "[Log] 639 cells for cell type cS28 (GREM1+ CAF)\n",
      "dimension of cell by gene count (639, 4618)\n",
      "[Log] 228 cells for cell type cS30 (CAF CCL8 Fibro-like)\n",
      "dimension of cell by gene count (228, 4618)\n",
      "[Log] 2255 cells for cell type cTNI16 (CD8+ CXCL13+ prolif)\n",
      "dimension of cell by gene count (2255, 4618)\n",
      "[Log] 20107 cells for cell type cP1 (Plasma IgA)\n",
      "dimension of cell by gene count (20107, 4618)\n",
      "[Log] 29726 cells for cell type Tumor cE03 (Stem/TA-like prolif)\n",
      "dimension of cell by gene count (29726, 4618)\n",
      "[Log] 5966 cells for cell type cTNI04 (CD4+ IL7R+CCL5+)\n",
      "dimension of cell by gene count (5966, 4618)\n",
      "[Log] 2832 cells for cell type cE06 (Immature Goblet)\n",
      "dimension of cell by gene count (2832, 4618)\n",
      "[Log] 1200 cells for cell type cS02 (Endo capillary)\n",
      "dimension of cell by gene count (1200, 4618)\n",
      "[Log] 383 cells for cell type cS24 (Fibro BMP-producing)\n",
      "dimension of cell by gene count (383, 4618)\n",
      "[Log] 473 cells for cell type Tumor cE08 (Goblet)\n",
      "dimension of cell by gene count (473, 4618)\n",
      "[Log] 4141 cells for cell type cE09 (Best4)\n",
      "dimension of cell by gene count (4141, 4618)\n",
      "[Log] 840 cells for cell type cM07 (pDC)\n",
      "dimension of cell by gene count (840, 4618)\n",
      "[Log] 1477 cells for cell type cTNI06 (CD4+ TFH)\n",
      "dimension of cell by gene count (1477, 4618)\n",
      "[Log] 2803 cells for cell type cTNI05 (CD4+ IL17+)\n",
      "dimension of cell by gene count (2803, 4618)\n",
      "[Log] 3325 cells for cell type cE03 (Stem/TA-like prolif)\n",
      "dimension of cell by gene count (3325, 4618)\n",
      "[Log] 457 cells for cell type cS31 (CAF stem niche Fibro-like)\n",
      "dimension of cell by gene count (457, 4618)\n",
      "[Log] 627 cells for cell type cS12 (Endo)\n",
      "dimension of cell by gene count (627, 4618)\n",
      "[Log] 3671 cells for cell type cMA01 (Mast)\n",
      "dimension of cell by gene count (3671, 4618)\n",
      "[Log] 254 cells for cell type cS33 (Schwann)\n",
      "dimension of cell by gene count (254, 4618)\n",
      "[Log] 1117 cells for cell type cTNI24 (NK GZMK+)\n",
      "dimension of cell by gene count (1117, 4618)\n",
      "[Log] 643 cells for cell type cTNI25 (NK XCL1+)\n",
      "dimension of cell by gene count (643, 4618)\n",
      "[Log] 2682 cells for cell type cTNI10 (CD8+ IL7R+)\n",
      "dimension of cell by gene count (2682, 4618)\n",
      "[Log] 936 cells for cell type cE10 (Tuft)\n",
      "dimension of cell by gene count (936, 4618)\n",
      "[Log] 252 cells for cell type cS15 (Pericyte)\n",
      "dimension of cell by gene count (252, 4618)\n",
      "[Log] 2708 cells for cell type cE02 (Stem/TA-like/Immature Goblet)\n",
      "dimension of cell by gene count (2708, 4618)\n",
      "[Log] 800 cells for cell type cS32 (Smooth Muscle)\n",
      "dimension of cell by gene count (800, 4618)\n",
      "[Log] 369 cells for cell type cTNI19 (gd-like T prolif)\n",
      "dimension of cell by gene count (369, 4618)\n",
      "[Log] 203 cells for cell type Tumor cE09 (Best4)\n",
      "dimension of cell by gene count (203, 4618)\n",
      "[Log] 688 cells for cell type cS22 (Fibro stem cell niche)\n",
      "dimension of cell by gene count (688, 4618)\n",
      "[Log] 3920 cells for cell type cTNI20 (PLZF+ T)\n",
      "dimension of cell by gene count (3920, 4618)\n",
      "[Log] 36490 cells for cell type Tumor cE01 (Stem/TA-like)\n",
      "dimension of cell by gene count (36490, 4618)\n",
      "[Log] 3878 cells for cell type cTNI15 (CD8+ CXCL13+ HSP+)\n",
      "dimension of cell by gene count (3878, 4618)\n",
      "[Log] 960 cells for cell type Tumor cE04 (Enterocyte 1)\n",
      "dimension of cell by gene count (960, 4618)\n",
      "[Log] 2003 cells for cell type cE04 (Enterocyte 1)\n",
      "dimension of cell by gene count (2003, 4618)\n",
      "[Log] 3178 cells for cell type cTNI18 (gd-like T PDCD1+)\n",
      "dimension of cell by gene count (3178, 4618)\n",
      "[Log] 6939 cells for cell type cB1 (B IGD+IgM+)\n",
      "dimension of cell by gene count (6939, 4618)\n",
      "[Log] 7886 cells for cell type cE01 (Stem/TA-like)\n",
      "dimension of cell by gene count (7886, 4618)\n",
      "[Log] 1051 cells for cell type cP3 (Plasma IgG prolif)\n",
      "dimension of cell by gene count (1051, 4618)\n"
     ]
    }
   ],
   "source": [
    "adata_markergs = adata[:, genemeanpd_filtered.index]\n",
    "genemeanpd_markergs, genemeanpdfc_markergs = gene_lambda_by_ct(adata_markergs, ctcol = 'group')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7dc4282-c4e3-4d29-85ff-48ed21c04609",
   "metadata": {},
   "source": [
    "# Step 3: Save the gene expression lambdas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "f48250cf-bedf-4c33-8c40-d2f0beb5617e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#outfile = '../gene_lambda_file.txt'\n",
    "genemeanpd_markergs.to_csv(outfile,sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "87194766-6537-4b9e-bb11-70aaee0f2ae1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cM03 (DC1)</th>\n",
       "      <th>cB2 (B GC-like)</th>\n",
       "      <th>cS05 (Endo venous)</th>\n",
       "      <th>cP2 (Plasma IgG)</th>\n",
       "      <th>cTNI12 (CD8+ IL7R+)</th>\n",
       "      <th>cS10 (Endo tip cells)</th>\n",
       "      <th>cM08 (AS-DC)</th>\n",
       "      <th>Tumor cE05 (Enterocyte 2)</th>\n",
       "      <th>cS21 (Fibro stem cell niche)</th>\n",
       "      <th>cTNI22 (cTNI22)</th>\n",
       "      <th>...</th>\n",
       "      <th>cS22 (Fibro stem cell niche)</th>\n",
       "      <th>cTNI20 (PLZF+ T)</th>\n",
       "      <th>Tumor cE01 (Stem/TA-like)</th>\n",
       "      <th>cTNI15 (CD8+ CXCL13+ HSP+)</th>\n",
       "      <th>Tumor cE04 (Enterocyte 1)</th>\n",
       "      <th>cE04 (Enterocyte 1)</th>\n",
       "      <th>cTNI18 (gd-like T PDCD1+)</th>\n",
       "      <th>cB1 (B IGD+IgM+)</th>\n",
       "      <th>cE01 (Stem/TA-like)</th>\n",
       "      <th>cP3 (Plasma IgG prolif)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SAMD11</th>\n",
       "      <td>0.000002</td>\n",
       "      <td>1.354575e-07</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>1.201868e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.414196e-07</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.471183e-05</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>3.665324e-07</td>\n",
       "      <td>5.086971e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.969258e-08</td>\n",
       "      <td>7.679331e-07</td>\n",
       "      <td>1.571570e-07</td>\n",
       "      <td>9.711975e-07</td>\n",
       "      <td>4.587755e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HES4</th>\n",
       "      <td>0.000013</td>\n",
       "      <td>1.165777e-05</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>4.646142e-06</td>\n",
       "      <td>3.429627e-05</td>\n",
       "      <td>9.395580e-05</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>1.226572e-04</td>\n",
       "      <td>1.098259e-05</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>9.994381e-05</td>\n",
       "      <td>6.711329e-05</td>\n",
       "      <td>1.021592e-05</td>\n",
       "      <td>1.102458e-04</td>\n",
       "      <td>4.320694e-05</td>\n",
       "      <td>1.901902e-05</td>\n",
       "      <td>9.532449e-06</td>\n",
       "      <td>3.761882e-05</td>\n",
       "      <td>6.681818e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISG15</th>\n",
       "      <td>0.000371</td>\n",
       "      <td>2.396111e-04</td>\n",
       "      <td>0.000769</td>\n",
       "      <td>1.925330e-04</td>\n",
       "      <td>2.311966e-04</td>\n",
       "      <td>4.572697e-04</td>\n",
       "      <td>0.000529</td>\n",
       "      <td>4.971079e-04</td>\n",
       "      <td>1.773285e-04</td>\n",
       "      <td>0.000452</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000202</td>\n",
       "      <td>6.084152e-04</td>\n",
       "      <td>6.048460e-04</td>\n",
       "      <td>8.492946e-04</td>\n",
       "      <td>7.696574e-03</td>\n",
       "      <td>6.667285e-03</td>\n",
       "      <td>7.051271e-04</td>\n",
       "      <td>2.246783e-04</td>\n",
       "      <td>8.444898e-05</td>\n",
       "      <td>1.927991e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNFRSF18</th>\n",
       "      <td>0.000006</td>\n",
       "      <td>8.850565e-06</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>2.501304e-05</td>\n",
       "      <td>1.415558e-04</td>\n",
       "      <td>1.726362e-05</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>2.664117e-06</td>\n",
       "      <td>1.609288e-07</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>1.472609e-03</td>\n",
       "      <td>2.307296e-06</td>\n",
       "      <td>5.548547e-04</td>\n",
       "      <td>5.822551e-07</td>\n",
       "      <td>1.339242e-06</td>\n",
       "      <td>5.439643e-04</td>\n",
       "      <td>3.157481e-06</td>\n",
       "      <td>1.783881e-06</td>\n",
       "      <td>2.542230e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNFRSF4</th>\n",
       "      <td>0.000006</td>\n",
       "      <td>6.559940e-06</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>1.902282e-05</td>\n",
       "      <td>3.706697e-05</td>\n",
       "      <td>9.563515e-04</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>2.876440e-06</td>\n",
       "      <td>2.534074e-06</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>5.529597e-05</td>\n",
       "      <td>2.146784e-06</td>\n",
       "      <td>1.335309e-04</td>\n",
       "      <td>2.988338e-06</td>\n",
       "      <td>1.819685e-07</td>\n",
       "      <td>3.784740e-05</td>\n",
       "      <td>2.829275e-06</td>\n",
       "      <td>1.436634e-06</td>\n",
       "      <td>1.265869e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP6AP1</th>\n",
       "      <td>0.000092</td>\n",
       "      <td>4.578629e-05</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>3.338679e-05</td>\n",
       "      <td>8.206867e-05</td>\n",
       "      <td>1.854192e-04</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>8.972633e-05</td>\n",
       "      <td>8.903323e-05</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>6.446808e-05</td>\n",
       "      <td>1.308182e-04</td>\n",
       "      <td>8.931441e-05</td>\n",
       "      <td>1.345284e-04</td>\n",
       "      <td>7.459446e-05</td>\n",
       "      <td>5.469417e-05</td>\n",
       "      <td>4.041530e-05</td>\n",
       "      <td>8.255739e-05</td>\n",
       "      <td>5.271606e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM50A</th>\n",
       "      <td>0.000057</td>\n",
       "      <td>6.297009e-05</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>2.755123e-05</td>\n",
       "      <td>1.120739e-04</td>\n",
       "      <td>1.255818e-04</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>1.129033e-04</td>\n",
       "      <td>1.115468e-04</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>1.541112e-04</td>\n",
       "      <td>1.449286e-04</td>\n",
       "      <td>1.354185e-04</td>\n",
       "      <td>9.213218e-05</td>\n",
       "      <td>5.820652e-05</td>\n",
       "      <td>1.227087e-04</td>\n",
       "      <td>5.150738e-05</td>\n",
       "      <td>9.278696e-05</td>\n",
       "      <td>5.626052e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LAGE3</th>\n",
       "      <td>0.000081</td>\n",
       "      <td>9.172346e-05</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>4.414699e-05</td>\n",
       "      <td>4.329798e-05</td>\n",
       "      <td>1.111914e-04</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>2.785378e-05</td>\n",
       "      <td>4.156548e-05</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>1.705586e-04</td>\n",
       "      <td>1.806062e-04</td>\n",
       "      <td>8.740751e-05</td>\n",
       "      <td>3.099945e-05</td>\n",
       "      <td>1.530810e-05</td>\n",
       "      <td>1.374337e-04</td>\n",
       "      <td>8.629313e-05</td>\n",
       "      <td>1.143672e-04</td>\n",
       "      <td>1.138378e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F8</th>\n",
       "      <td>0.000001</td>\n",
       "      <td>7.350558e-06</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>6.585625e-07</td>\n",
       "      <td>2.725360e-06</td>\n",
       "      <td>6.688461e-06</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.451038e-08</td>\n",
       "      <td>6.042331e-06</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>6.544941e-06</td>\n",
       "      <td>2.584643e-06</td>\n",
       "      <td>1.148462e-05</td>\n",
       "      <td>3.076564e-06</td>\n",
       "      <td>2.969258e-08</td>\n",
       "      <td>1.202523e-05</td>\n",
       "      <td>8.105909e-07</td>\n",
       "      <td>2.945273e-06</td>\n",
       "      <td>1.505940e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLIC2</th>\n",
       "      <td>0.000162</td>\n",
       "      <td>6.653794e-07</td>\n",
       "      <td>0.000224</td>\n",
       "      <td>6.083002e-07</td>\n",
       "      <td>5.775506e-07</td>\n",
       "      <td>6.426967e-05</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>5.627824e-07</td>\n",
       "      <td>7.764958e-05</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>1.473982e-07</td>\n",
       "      <td>1.790717e-06</td>\n",
       "      <td>2.467462e-07</td>\n",
       "      <td>1.533624e-06</td>\n",
       "      <td>3.053337e-08</td>\n",
       "      <td>3.632715e-07</td>\n",
       "      <td>7.814851e-07</td>\n",
       "      <td>1.573451e-06</td>\n",
       "      <td>8.360396e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4618 rows Ã— 98 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cM03 (DC1)  cB2 (B GC-like)  cS05 (Endo venous)  cP2 (Plasma IgG)  \\\n",
       "SAMD11      0.000002     1.354575e-07            0.000008      1.201868e-06   \n",
       "HES4        0.000013     1.165777e-05            0.000024      4.646142e-06   \n",
       "ISG15       0.000371     2.396111e-04            0.000769      1.925330e-04   \n",
       "TNFRSF18    0.000006     8.850565e-06            0.000007      2.501304e-05   \n",
       "TNFRSF4     0.000006     6.559940e-06            0.000058      1.902282e-05   \n",
       "...              ...              ...                 ...               ...   \n",
       "ATP6AP1     0.000092     4.578629e-05            0.000107      3.338679e-05   \n",
       "FAM50A      0.000057     6.297009e-05            0.000098      2.755123e-05   \n",
       "LAGE3       0.000081     9.172346e-05            0.000126      4.414699e-05   \n",
       "F8          0.000001     7.350558e-06            0.000035      6.585625e-07   \n",
       "CLIC2       0.000162     6.653794e-07            0.000224      6.083002e-07   \n",
       "\n",
       "          cTNI12 (CD8+ IL7R+)  cS10 (Endo tip cells)  cM08 (AS-DC)  \\\n",
       "SAMD11           0.000000e+00           2.414196e-07      0.000004   \n",
       "HES4             3.429627e-05           9.395580e-05      0.000052   \n",
       "ISG15            2.311966e-04           4.572697e-04      0.000529   \n",
       "TNFRSF18         1.415558e-04           1.726362e-05      0.000011   \n",
       "TNFRSF4          3.706697e-05           9.563515e-04      0.000051   \n",
       "...                       ...                    ...           ...   \n",
       "ATP6AP1          8.206867e-05           1.854192e-04      0.000216   \n",
       "FAM50A           1.120739e-04           1.255818e-04      0.000125   \n",
       "LAGE3            4.329798e-05           1.111914e-04      0.000143   \n",
       "F8               2.725360e-06           6.688461e-06      0.000000   \n",
       "CLIC2            5.775506e-07           6.426967e-05      0.000208   \n",
       "\n",
       "          Tumor cE05 (Enterocyte 2)  cS21 (Fibro stem cell niche)  \\\n",
       "SAMD11                 0.000000e+00                  1.471183e-05   \n",
       "HES4                   1.226572e-04                  1.098259e-05   \n",
       "ISG15                  4.971079e-04                  1.773285e-04   \n",
       "TNFRSF18               2.664117e-06                  1.609288e-07   \n",
       "TNFRSF4                2.876440e-06                  2.534074e-06   \n",
       "...                             ...                           ...   \n",
       "ATP6AP1                8.972633e-05                  8.903323e-05   \n",
       "FAM50A                 1.129033e-04                  1.115468e-04   \n",
       "LAGE3                  2.785378e-05                  4.156548e-05   \n",
       "F8                     3.451038e-08                  6.042331e-06   \n",
       "CLIC2                  5.627824e-07                  7.764958e-05   \n",
       "\n",
       "          cTNI22 (cTNI22)  ...  cS22 (Fibro stem cell niche)  \\\n",
       "SAMD11           0.000000  ...                      0.000039   \n",
       "HES4             0.000009  ...                      0.000016   \n",
       "ISG15            0.000452  ...                      0.000202   \n",
       "TNFRSF18         0.000084  ...                      0.000002   \n",
       "TNFRSF4          0.000062  ...                      0.000002   \n",
       "...                   ...  ...                           ...   \n",
       "ATP6AP1          0.000043  ...                      0.000094   \n",
       "FAM50A           0.000110  ...                      0.000091   \n",
       "LAGE3            0.000099  ...                      0.000049   \n",
       "F8               0.000012  ...                      0.000004   \n",
       "CLIC2            0.000001  ...                      0.000093   \n",
       "\n",
       "          cTNI20 (PLZF+ T)  Tumor cE01 (Stem/TA-like)  \\\n",
       "SAMD11        3.665324e-07               5.086971e-07   \n",
       "HES4          9.994381e-05               6.711329e-05   \n",
       "ISG15         6.084152e-04               6.048460e-04   \n",
       "TNFRSF18      1.472609e-03               2.307296e-06   \n",
       "TNFRSF4       5.529597e-05               2.146784e-06   \n",
       "...                    ...                        ...   \n",
       "ATP6AP1       6.446808e-05               1.308182e-04   \n",
       "FAM50A        1.541112e-04               1.449286e-04   \n",
       "LAGE3         1.705586e-04               1.806062e-04   \n",
       "F8            6.544941e-06               2.584643e-06   \n",
       "CLIC2         1.473982e-07               1.790717e-06   \n",
       "\n",
       "          cTNI15 (CD8+ CXCL13+ HSP+)  Tumor cE04 (Enterocyte 1)  \\\n",
       "SAMD11                  0.000000e+00               0.000000e+00   \n",
       "HES4                    1.021592e-05               1.102458e-04   \n",
       "ISG15                   8.492946e-04               7.696574e-03   \n",
       "TNFRSF18                5.548547e-04               5.822551e-07   \n",
       "TNFRSF4                 1.335309e-04               2.988338e-06   \n",
       "...                              ...                        ...   \n",
       "ATP6AP1                 8.931441e-05               1.345284e-04   \n",
       "FAM50A                  1.354185e-04               9.213218e-05   \n",
       "LAGE3                   8.740751e-05               3.099945e-05   \n",
       "F8                      1.148462e-05               3.076564e-06   \n",
       "CLIC2                   2.467462e-07               1.533624e-06   \n",
       "\n",
       "          cE04 (Enterocyte 1)  cTNI18 (gd-like T PDCD1+)  cB1 (B IGD+IgM+)  \\\n",
       "SAMD11           2.969258e-08               7.679331e-07      1.571570e-07   \n",
       "HES4             4.320694e-05               1.901902e-05      9.532449e-06   \n",
       "ISG15            6.667285e-03               7.051271e-04      2.246783e-04   \n",
       "TNFRSF18         1.339242e-06               5.439643e-04      3.157481e-06   \n",
       "TNFRSF4          1.819685e-07               3.784740e-05      2.829275e-06   \n",
       "...                       ...                        ...               ...   \n",
       "ATP6AP1          7.459446e-05               5.469417e-05      4.041530e-05   \n",
       "FAM50A           5.820652e-05               1.227087e-04      5.150738e-05   \n",
       "LAGE3            1.530810e-05               1.374337e-04      8.629313e-05   \n",
       "F8               2.969258e-08               1.202523e-05      8.105909e-07   \n",
       "CLIC2            3.053337e-08               3.632715e-07      7.814851e-07   \n",
       "\n",
       "          cE01 (Stem/TA-like)  cP3 (Plasma IgG prolif)  \n",
       "SAMD11           9.711975e-07             4.587755e-07  \n",
       "HES4             3.761882e-05             6.681818e-06  \n",
       "ISG15            8.444898e-05             1.927991e-04  \n",
       "TNFRSF18         1.783881e-06             2.542230e-05  \n",
       "TNFRSF4          1.436634e-06             1.265869e-05  \n",
       "...                       ...                      ...  \n",
       "ATP6AP1          8.255739e-05             5.271606e-05  \n",
       "FAM50A           9.278696e-05             5.626052e-05  \n",
       "LAGE3            1.143672e-04             1.138378e-04  \n",
       "F8               2.945273e-06             1.505940e-06  \n",
       "CLIC2            1.573451e-06             8.360396e-08  \n",
       "\n",
       "[4618 rows x 98 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genemeanpd_markergs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c61024bc-7de6-48f7-9614-487bfe8278be",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "stvit-cpu",
   "language": "python",
   "name": "stvit-cpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
